### PR TITLE
fix(dgw): forward protocol advertised by mDNS scanner

### PR DESCRIPTION
**Pull request by irvingoujAtDevolution**
Current implementation only guesses what service type it is by using port, mdns allows more than just port guessing, enable passing kown protocol types to caller, for unknwon protocols, the caller can keep guess what it is using port.

### CODE REVIEW

File: devolutions-gateway/src/api/network_scan.rs [hunk 1]
```diff
@@ -115,23 +115,49 @@ pub struct NetworkScanResponse {
...
+                5986 => ApplicationProtocol::Known(Protocol::WinrmHttpsPwsh),
+                _ => ApplicationProtocol::unknown(),
+            }
         };
 
         Self { ip, hostname, protocol }
     }
 }
+
+fn network_scanner_protocol_to_gateway_protocol(protocol: Option<scanner::Protocol>) -> ApplicationProtocol {
```
Comment (CBenoit) [inline devolutions-gateway/src/api/network_scan.rs:143]: suggestion: Modify this function to take a `scanner::Protocol` directly.

```suggestion
fn network_scanner_protocol_to_gateway_protocol(protocol: scanner::Protocol) -> ApplicationProtocol {
```

File: devolutions-gateway/src/api/network_scan.rs [hunk 2]
```diff
@@ -115,23 +115,49 @@ pub struct NetworkScanResponse {
...
-            443 => ApplicationProtocol::Known(Protocol::Https),
-            389 => ApplicationProtocol::Known(Protocol::Ldap),
-            636 => ApplicationProtocol::Known(Protocol::Ldaps),
-            3389 => ApplicationProtocol::Known(Protocol::Rdp),
-            5900 => ApplicationProtocol::Known(Protocol::Vnc),
-            5985 => ApplicationProtocol::Known(Protocol::WinrmHttpPwsh),
-            5986 => ApplicationProtocol::Known(Protocol::WinrmHttpsPwsh),
-            _ => ApplicationProtocol::unknown(),
+        let protocol = if service.is_some() {
+            network_scanner_protocol_to_gateway_protocol(service)
```
Comment (CBenoit) [inline devolutions-gateway/src/api/network_scan.rs:122]: suggestion: And retrieve a `scanner::Protocol` using pattern matching.

```suggestion
        let protocol = if let Some(protocol) = service {
            network_scanner_protocol_to_gateway_protocol(protocol)
```

Comment (CBenoit) [inline devolutions-gateway/src/api/network_scan.rs:122]: suggestion: You can ignore this if you don’t agree, but I feel we could just inline the `network_scanner_protocol_to_gateway_protocol` here.

File: crates/network-scanner/src/scanner.rs [hunk 3]
```diff
@@ -351,3 +349,39 @@ pub struct NetworkScannerParams {
...
+    /// Hypertext Transfer Protocol
+    Http,
+    /// Hypertext Transfer Protocol Secure
+    Https,
+    /// LDAP Protocol
+    Ldap,
+    /// Secure LDAP Protocol
+    Ldaps,
+    /// Devolutions Gateway Tunnel (generic JMUX tunnel)
+    Tunnel,
```
Comment (CBenoit) [inline crates/network-scanner/src/scanner.rs:386]: issue: The "Devolutions Gateway Tunnel" aka "JMUX" protocol is not something we can scan for. You can remove it from the enum.

File: crates/network-scanner/src/scanner.rs [hunk 4]
```diff
@@ -262,21 +262,19 @@ impl NetworkScanner {
     }
 }
 
-type ResultReceiver = tokio::sync::mpsc::Receiver<(IpAddr, Option<String>, u16)>;
+pub type ScanEntry = (IpAddr, Option<String>, u16, Option<Protocol>);
+type ResultReceiver = tokio::sync::mpsc::Receiver<ScanEntry>;
 pub struct NetworkScannerStream {
```
Comment (CBenoit) [inline crates/network-scanner/src/scanner.rs:267]: style: You can consider adding a few newlines to make the code more readable. This is a general comment with this codebase, although it’s not critical.

File: crates/network-scanner/src/scanner.rs [hunk 5]
```diff
@@ -351,3 +349,39 @@ pub struct NetworkScannerParams {
     pub mdns_single_query_timeout: u64, // in milliseconds
     pub max_wait_time: u64,             // max_wait for entire scan duration in milliseconds, suggested!
 }
+
+#[derive(Debug, Clone)]
+pub enum Protocol {
+    /// Wayk Remote Desktop Protocol
+    Wayk,
```
Comment (CBenoit) [inline crates/network-scanner/src/scanner.rs:356]: suggestion: I think you can safely remove Wayk from this list too. The associated product is dead.

File: crates/network-scanner/src/mdns.rs [hunk 6]
```diff
@@ -110,3 +115,35 @@ pub fn mdns_query_scan(
...
+        "_ssh._tcp" => Some(Protocol::Ssh),
+        "_sftp._tcp" => Some(Protocol::Sftp),
+        "_scp._tcp" => Some(Protocol::Scp),
+        "_telnet._tcp" => Some(Protocol::Telnet),
+        "_ldap._tcp" => Some(Protocol::Ldap),
+        "_ldaps._tcp" => Some(Protocol::Ldaps),
+        // https://jonathanmumm.com/tech-it/mdns-bonjour-bible-common-service-strings-for-various-vendors/
+        // OSX Screen Sharing
+        "_rfb._tcp" => Some(Protocol::Ard),
+        // Note: RDP, VNC, Wayk, and SSH-based PowerShell (SshPwsh), as well as tunneling (Tunnel), are not standardized service types
```
Comment (CBenoit) [inline crates/network-scanner/src/mdns.rs:144]: nitpick: JMUX (Devolutions Gateway Tunnel) is not worth mentioning here.

```suggestion
        // Note: RDP, VNC, Wayk, and SSH-based PowerShell (SshPwsh) are not standardized service types
```

File: crates/network-scanner/src/scanner.rs (hunk 3 above)
Comment (irvingoujAtDevolution) [reply crates/network-scanner/src/scanner.rs]: updated

File: devolutions-gateway/src/api/network_scan.rs (hunk 2 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/api/network_scan.rs]: updated

File: crates/network-scanner/src/scanner.rs (hunk 4 above)
Comment (irvingoujAtDevolution) [reply crates/network-scanner/src/scanner.rs]: updated

File: crates/network-scanner/src/scanner.rs (hunk 5 above)
Comment (irvingoujAtDevolution) [reply crates/network-scanner/src/scanner.rs]: updated, removed non-standard protocols

File: crates/network-scanner/src/scanner.rs [hunk 7]
```diff
@@ -276,21 +276,23 @@ impl NetworkScanner {
     }
 }
 
-type ResultReceiver = tokio::sync::mpsc::Receiver<(IpAddr, Option<String>, u16)>;
+/// ScanEntry is a tuple of (IpAddr, Option<String>, u16, Option<Protocol>)
+/// where IpAddr is the ip address of the device
+/// Option<String> is the hostname of the device
+/// u16 is the port number
+/// Option<Protocol> is the protocol/Service running on the port
+pub type ScanEntry = (IpAddr, Option<String>, u16, Option<Protocol>);
```
Comment (CBenoit) [inline crates/network-scanner/src/scanner.rs:284]: suggestion (non-blocking): Use a struct for readability.

File: devolutions-gateway/src/api/network_scan.rs [hunk 8]
```diff
@@ -33,10 +33,10 @@ pub async fn handler(
         loop {
             tokio::select! {
                 result = stream.recv() => {
-                    let Some((ip, dns, port)) = result else{
+                    let Some((ip, dns, port,protocol)) = result else{
```
Comment (CBenoit) [inline devolutions-gateway/src/api/network_scan.rs:36]: style: Spacing

**Review by CBenoit — APPROVED** [review]
LGTM!

**Pull request merged**